<a href="https://colab.research.google.com/github/SatyaIyerRaghav/PandaPractisePrograms/blob/main/Practice%2BExercise%2B2%2BMovies.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to download a movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [1]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [2]:
# Write your code for importing the csv file here
movies = pd.read_csv('Movies.csv')
movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
3848  Color      Shane Carruth                   143.0      77.0   
3849  Color   Neill Dela Llana                    35.0      80.0   
3850  Color   Robert Rodriguez                    56.0      81.0   
3851  Color       Edward Burns                    14.0      95.0   
3852  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                         0.0                   855.0    Joel David Moore   
1                       563.0                  1000.0       Orlando Bloom   
2                         0.0                   161.0        Rory Kinnear   
3                     22000.0                 23000.0      Christian Bale   
4                       475.0                   530.0     Samantha Morton   
...                       ...                     ...                 ...   
3848                    291.0                     8.0      David Sullivan   
3849                      0.0                     0.0     Edgar Tancangco   
3850                      0.0                     6.0     Peter Marquardt   
3851                      0.0                   133.0  Caitlin FitzGerald   
3852                     16.0                    16.0    Brian Herzlinger   

      actor_1_facebook_likes        gross  \
0                     1000.0  760505847.0   
1                    40000.0  309404152.0   
2                    11000.0  200074175.0   
3                    27000.0  448130642.0   
4                      640.0   73058679.0   
...                      ...          ...   
3848                   291.0     424760.0   
3849                     0.0      70071.0   
3850                   121.0    2040920.0   
3851                   296.0       4584.0   
3852                    86.0      85222.0   

                                   genres  ... num_user_for_reviews language  \
0         Action|Adventure|Fantasy|Sci-Fi  ...               3054.0  English   
1                Action|Adventure|Fantasy  ...               1238.0  English   
2               Action|Adventure|Thriller  ...                994.0  English   
3                         Action|Thriller  ...               2701.0  English   
4                 Action|Adventure|Sci-Fi  ...                738.0  English   
...                                   ...  ...                  ...      ...   
3848                Drama|Sci-Fi|Thriller  ...                371.0  English   
3849                             Thriller  ...                 35.0  English   
3850  Action|Crime|Drama|Romance|Thriller  ...                130.0  Spanish   
3851                         Comedy|Drama  ...                 14.0  English   
3852                          Documentary  ...                 84.0  English   

          country  content_rating       budget  title_year  \
0             USA           PG-13  237000000.0      2009.0   
1             USA           PG-13  300000000.0      2007.0   
2              UK           PG-13  245000000.0      2015.0   
3             USA           PG-13  250000000.0      2012.0   
4             USA           PG-13  263700000.0      2012.0   
...           ...             ...          ...         ...   
3848          USA           PG-13       7000.0      2004.0   
3849  Philippines       Not Rated       7000.0      2005.0   
3850          USA               R       7000.0      1992.0   
3851          USA       Not Rated       9000.0      2011.0   
3852          USA              PG       1100.0      2004.0   

     actor_2_face

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [3]:
# Write your code for inspection here
print(movies)

      color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
3848  Color      Shane Carruth                   143.0      77.0   
3849  Color   Neill Dela Llana                    35.0      80.0   
3850  Color   Robert Rodriguez                    56.0      81.0   
3851  Color       Edward Burns                    14.0      95.0   
3852  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                         0.0                   855.0    Joel David Moore   
1                       563.0

#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

In [18]:
m = (movies.isna()=="True")
print(m.sum())

color                        0
director_name                0
num_critic_for_reviews       0
duration                     0
director_facebook_likes      0
actor_3_facebook_likes       0
actor_2_name                 0
actor_1_facebook_likes       0
gross                        0
genres                       0
actor_1_name                 0
movie_title                  0
num_voted_users              0
cast_total_facebook_likes    0
actor_3_name                 0
facenumber_in_poster         0
plot_keywords                0
movie_imdb_link              0
num_user_for_reviews         0
language                     0
country                      0
content_rating               0
budget                       0
title_year                   0
actor_2_facebook_likes       0
imdb_score                   0
aspect_ratio                 0
movie_facebook_likes         0
dtype: int64


#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [23]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations
# #  color
# -  director_facebook_likes
# -  actor_1_facebook_likes
# -  actor_2_facebook_likes
# -  actor_3_facebook_likes
# -  actor_2_name
# -  cast_total_facebook_likes
# -  actor_3_name
# -  duration
# -  facenumber_in_poster
# -  content_rating
# -  country
# -  movie_imdb_link
# -  aspect_ratio
# -  plot_keywords
m = movies.drop(columns = ['color','director_facebook_likes','actor_1_facebook_likes','actor_2_facebook_likes','actor_3_facebook_likes','actor_2_name','cast_total_facebook_likes','actor_3_name','duration','facenumber_in_poster','content_rating','country','movie_imdb_link','aspect_ratio','plot_keywords'])
print(m.head())

       director_name  num_critic_for_reviews        gross  \
0      James Cameron                   723.0  760505847.0   
1     Gore Verbinski                   302.0  309404152.0   
2         Sam Mendes                   602.0  200074175.0   
3  Christopher Nolan                   813.0  448130642.0   
4     Andrew Stanton                   462.0   73058679.0   

                            genres     actor_1_name  \
0  Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1         Action|Adventure|Fantasy      Johnny Depp   
2        Action|Adventure|Thriller  Christoph Waltz   
3                  Action|Thriller        Tom Hardy   
4          Action|Adventure|Sci-Fi     Daryl Sabara   

                                 movie_title  num_voted_users  \
0                                    Avatar            886204   
1  Pirates of the Caribbean: At World's End            471220   
2                                   Spectre            275868   
3                     The Dark Knight Rise

#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'. 

In [26]:
# Write you code here
k = m.isnull().sum()
print(k)

director_name             0
num_critic_for_reviews    1
gross                     0
genres                    0
actor_1_name              0
movie_title               0
num_voted_users           0
num_user_for_reviews      0
language                  3
budget                    0
title_year                0
imdb_score                0
movie_facebook_likes      0
dtype: int64


#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [55]:
# Write your code for filling the NaN values in the 'language' column here
k = m.fillna('English')
print((movies.language == 'English').sum())


3674


#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [52]:
# Write your code for unit conversion here
movies['budget'] = movies['budget']/1000000
movies['gross']=movies['gross']/1000000
print(movies)

      color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
3848  Color      Shane Carruth                   143.0      77.0   
3849  Color   Neill Dela Llana                    35.0      80.0   
3850  Color   Robert Rodriguez                    56.0      81.0   
3851  Color       Edward Burns                    14.0      95.0   
3852  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                         0.0                   855.0    Joel David Moore   
1                       563.0

**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [56]:
# Write your code for creating the profit column here
movies['profit']=movies['gross']-movies['budget']

In [65]:
# Write your code for sorting the dataframe here
q = movies.sort_values('profit',ascending=False)
print(q['profit'])

0       0.000524
28      0.000502
25      0.000459
2704    0.000450
2748    0.000424
          ...   
2147   -0.002127
2136   -0.002398
2693   -0.002500
3282   -0.004200
2678   -0.012213
Name: profit, Length: 3853, dtype: float64


In [66]:
# top10 = # Write your code to get the top 10 profiting movies here
top10 = q.head(10)
top10

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
28    Color    Colin Trevorrow                   644.0     124.0   
25    Color      James Cameron                   315.0     194.0   
2704  Color       George Lucas                   282.0     125.0   
2748  Color   Steven Spielberg                   215.0     120.0   
16    Color        Joss Whedon                   703.0     173.0   
482   Color       Roger Allers                   186.0      73.0   
230   Color       George Lucas                   320.0     136.0   
64    Color  Christopher Nolan                   645.0     152.0   
419   Color          Gary Ross                   673.0     142.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
0                         0.0                   855.0   Joel David Moore   
28                      365.0                  1000.0         Judy Greer   
25                        0.0                   794.0       Kate Winslet   
2704                      0.0                   504.0      Peter Cushing   
2748                  14000.0                   548.0        Dee Wallace   
16                        0.0                 19000.0  Robert Downey Jr.   
482                      28.0                   847.0        Nathan Lane   
230                       0.0                  1000.0        Liam Neeson   
64                    22000.0                 11000.0       Heath Ledger   
419                     378.0                   575.0    Josh Hutcherson   

      actor_1_facebook_likes     gross  \
0                     1000.0  0.000761   
28                    3000.0  0.000652   
25                   29000.0  0.000659   
2704                 11000.0  0.000461   
2748                   861.0  0.000435   
16                   26000.0  0.000623   
482                   2000.0  0.000423   
230                  20000.0  0.000475   
64                   23000.0  0.000533   
419                  34000.0  0.000408   

                                        genres  ... language country  \
0              Action|Adventure|Fantasy|Sci-Fi  ...  English     USA   
28            Action|Adventure|Sci-Fi|Thriller  ...  English     USA   
25                               Drama|Romance  ...  English     USA   
2704           Action|Adventure|Fantasy|Sci-Fi  ...  English     USA   
2748                             Family|Sci-Fi  ...  English     USA   
16                     Action|Adventure|Sci-Fi  ...  English     USA   
482   Adventure|Animation|Drama|Family|Musical  ...  English     USA   
230            Action|Adventure|Fantasy|Sci-Fi  ...  English     USA   
64                 Action|Crime|Drama|Thriller  ...  English     USA   
419            Adventure|Drama|Sci-Fi|Thriller  ...  English     USA   

      content_rating    budget title_year  actor_2_facebook_likes imdb_score  \
0              PG-13  0.000237     2009.0                   936.0        7.9   
28             PG-13  0.000150     2015.0                  2000.0        7.0   
25             PG-13  0.000200     1997.0                 14000.0        7.7   
2704              PG  0.000011     1977.0                  1000.0        8.7   
2748              PG  0.000010     1982.0                   725.0        7.9   
16             PG-13  0.000220     2012.0                 21000.0        8.1   
482                G  0.000045     1994.0                   886.0        8.5   
230               PG  0.000115     1999.0                 14000.0        6.5   
64             PG-13  0.000185     2008.0                 13000.0        9.0   
419            PG-13  0.000078     2012.0                 14000.0        7.3   

     aspect_ratio  movie_facebook_likes    profit  
0            1.78                 33000  0.000524  
28           2.00                150000  0.000502  
25           2.35                 26000  0.000459  
2704         2.35                 33000  0.000450  
2748         1.85      

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

In [72]:
print(q['movie_title'])

0                                   Avatar 
28                          Jurassic World 
25                                 Titanic 
2704    Star Wars: Episode IV - A New Hope 
2748            E.T. the Extra-Terrestrial 
                       ...                 
2147                              Steamboy 
2136                     Princess Mononoke 
2693                              Fateless 
3282                        Lady Vengeance 
2678                              The Host 
Name: movie_title, Length: 3853, dtype: object


#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

**Subtask 3.3: Find IMDb Top 250**

Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000. 

Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [4]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe 
# and name that dataframe as 'IMDb_Top_250'

IMDb_Top_250 = movies.sort_values('imdb_score',ascending=False).rank()
n = IMDb_Top_250['num_voted_users']>25000
print(IMDb_Top_250)


       color  director_name  num_critic_for_reviews  duration  \
1795  1991.0         1035.5                  2714.5    3593.0   
3016  1991.0         1015.0                  2776.5    3784.5   
2543  1991.0         1015.0                  2131.0    3837.5   
64    1991.0          556.5                  3831.5    3692.0   
325   1991.0         2721.5                  3460.5    3813.5   
...      ...            ...                     ...       ...   
3161  1991.0         2791.5                   216.0     929.0   
3049  1991.0          877.5                   216.0     751.5   
2085  1991.0         1442.5                  1582.0     356.5   
2111  1991.0          315.0                   324.5     356.5   
2540  1991.0         1791.0                  1174.5    2621.0   

      director_facebook_likes  actor_3_facebook_likes  actor_2_name  \
1795                    336.5                  2072.5        1682.5   
3016                    336.5                  3718.5        2397.0   
2543  

#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_250*? </font>

In [10]:
def bucket(x):
#     x=imdb_score
    if x>=7.5 and x<8:
        return "5"
    elif x>=8 and x<8.5:
        return "4"
    elif x>=8.5 and x<9:
        return "3"
    elif x>=9 and x<9.5:
        return "2"
    elif x>=9.5 and x<10:
        return "1"
IMDb_Top_250["bucket"]=IMDb_Top_250["imdb_score"].apply(bucket)
IMDb_Top_250.groupby("bucket", axis=0).count()

color  director_name  num_critic_for_reviews  duration  \
bucket                                                           
4           1              1                       1         1   

        director_facebook_likes  actor_3_facebook_likes  actor_2_name  \
bucket                                                                  
4                             1                       1             1   

        actor_1_facebook_likes  gross  genres  ...  num_user_for_reviews  \
bucket                                         ...                         
4                            1      1       1  ...                     1   

        language  country  content_rating  budget  title_year  \
bucket                                                          
4              1        1               1       1           1   

        actor_2_facebook_likes  imdb_score  aspect_ratio  movie_facebook_likes  
bucket                                                                          
4                            1           1             1                     1  

[1 rows x 28 columns]

**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [ ]:
k = movies['actor_1_name']

In [ ]:
# Write your code for creating three new dataframes here
Meryl_Streep = movies[(movies['actor_1_name'] == 'Meryl Streep')]
print(Meryl_Streep)

In [ ]:
Leo_Caprio = movies[(movies['actor_1_name'] == 'Leonardo DiCaprio')]
print(Leo_Caprio)

In [ ]:
Brad_Pitt = movies[(movies['actor_1_name'] ==  'Brad Pitt' )]
print(Brad_Pitt)

In [ ]:
# Write your code for combining the three dataframes here
Combined = pd.concat([Meryl_Streep,Leo_Caprio,Brad_Pitt],axis=0)
print(Combined)

In [26]:
# Write your code for grouping the combined dataframe here
Combined.groupby(by =(movies['actor_1_name']) )

In [36]:
# Write the code for finding the mean of critic reviews and audience reviews here
p = Combined.groupby(by =(movies['actor_1_name']) ).mean()[['num_critic_for_reviews',"num_user_for_reviews"]]
print(p)

                   num_critic_for_reviews  num_user_for_reviews
actor_1_name                                                   
Brad Pitt                      245.000000            742.352941
Leonardo DiCaprio              330.190476            914.476190
Meryl Streep                   181.454545            297.181818


<ipython-input-36-1e872a15bcae>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  p = Combined.groupby(by =(movies['actor_1_name']) ).mean()[['num_critic_for_reviews',"num_user_for_reviews"]]


#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>